### PR TITLE
NMS-12729: OpenConfig streaming support

### PR DESCRIPTION
**Pull request by j-white**
JIRA: https://issues.opennms.org/browse/NMS-12729

### CODE REVIEW

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: container/features/src/main/resources/features-sentinel.xml [hunk 1]
```diff
@@ -166,7 +166,6 @@
         <bundle>mvn:org.opennms.features.distributed/org.opennms.features.distributed.collection/${project.version}</bundle>
     </feature>
 
-
```
Comment (j-white) [inline container/features/src/main/resources/features-sentinel.xml:169]: Unnecessary change.

File: container/karaf/src/main/filtered-resources/etc/org.apache.karaf.features.cfg [hunk 2]
```diff
@@ -105,6 +105,8 @@ featuresBoot = ( \
   opennms-telemetry-jti,\
   opennms-telemetry-nxos,\
   opennms-telemetry-graphite,\
+  opennms-telemetry-openconfig-client,\
```
Comment (j-white) [inline container/karaf/src/main/filtered-resources/etc/org.apache.karaf.features.cfg:108]: Wouldn't we expect `opennms-telemetry-openconfig` to also load `opennms-telemetry-openconfig-client`?

File: core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/Interpolator.java [hunk 3]
```diff
@@ -93,7 +93,24 @@ public static String interpolate(final String raw, final Scope scope) {
         outerMatcher.appendTail(stringBuffer);
         return stringBuffer.toString();
     }
-    
+
+    public static Optional<ContextKey> getContextKeyFromMateData(final String raw) {
```
Comment (j-white) [inline core/ipc/rpc/utils/src/main/java/org/opennms/core/rpc/utils/mate/Interpolator.java:97]: This still needed?

File: core/lib/src/main/java/org/opennms/core/utils/StringUtils.java [hunk 4]
```diff
@@ -453,4 +453,20 @@ public static String getHumanReadableByteCount(long bytes, boolean si) {
             return Optional.empty();
         }
     }
+
+    public static Integer parseInt(String value, Integer defaultValue) {
```
Comment (j-white) [inline core/lib/src/main/java/org/opennms/core/utils/StringUtils.java:457]: We should probably remove one of these.

File: features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/OpennmsKafkaProducer.java [hunk 5]
```diff
@@ -397,7 +397,7 @@ private void sendRecord(Callable<ProducerRecord<byte[], byte[]>> callable, Consu
         }
     }
 
-    private void processKafkaSendQueue() {
+    private void    processKafkaSendQueue() {
```
Comment (j-white) [inline features/kafka/producer/src/main/java/org/opennms/features/kafka/producer/OpennmsKafkaProducer.java:400]: Unnecessary change.

File: features/telemetry/itests/src/test/java/org/opennms/netmgt/telemetry/itests/OpenConfigAdapterIT.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.telemetry.itests;
+
+public class OpenConfigAdapterIT {
```
Comment (j-white) [inline features/telemetry/itests/src/test/java/org/opennms/netmgt/telemetry/itests/OpenConfigAdapterIT.java:31]: TODO: Implement or delete.

File: features/telemetry/listeners/src/main/resources/OSGI-INF/blueprint/blueprint.xml [hunk 7]
```diff
@@ -29,4 +29,5 @@
             <entry key="registration.export" value="true" />
         </service-properties>
     </service>
+
```
Comment (j-white) [inline features/telemetry/listeners/src/main/resources/OSGI-INF/blueprint/blueprint.xml:32]: Unnecessary change.

File: opennms-base-assembly/src/main/filtered/etc/telemetryd-configuration.xml [hunk 8]
```diff
@@ -24,6 +24,34 @@
         </adapter>
     </queue>
 
+    <!-- OpenConfig -->
+    <connector name="OpenConfig-Connector"
+               class-name="org.opennms.netmgt.telemetry.protocols.openconfig.connector.OpenConfigConnector"
+               service-name="OpenConfig"
+               queue="OpenConfig"
+               enabled="true">
+        <package name="OpenConfig-Default">
```
Comment (j-white) [inline opennms-base-assembly/src/main/filtered/etc/telemetryd-configuration.xml:33]: TODO: Improve this example.

File: opennms-dao/src/main/java/org/opennms/netmgt/dao/support/DefaultFilterWatcher.java [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+import org.opennms.netmgt.events.api.model.IEvent;
+import org.opennms.netmgt.filter.api.FilterDao;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * TODO: Review/document concurrency model
```
Comment (j-white) [inline opennms-dao/src/main/java/org/opennms/netmgt/dao/support/DefaultFilterWatcher.java:64]: TODO

File: opennms-dao/src/main/resources/META-INF/opennms/applicationContext-dao.xml [hunk 10]
```diff
@@ -169,4 +169,5 @@
     <constructor-arg value="org.opennms.netmgt.config.poller.PollerConfiguration" />
     <constructor-arg value="file:${opennms.home}/etc/poller-configuration.xml" />
   </bean>
+
```
Comment (j-white) [inline opennms-dao/src/main/resources/META-INF/opennms/applicationContext-dao.xml:172]: Unnecessary change.

File: opennms-dao/src/main/resources/META-INF/opennms/applicationContext-shared.xml [hunk 11]
```diff
@@ -289,4 +289,5 @@
     <onmsgi:list id="alarmEntityListeners" interface="org.opennms.netmgt.dao.api.AlarmEntityListener">
         <onmsgi:listener ref="alarmEntityNotifier" bind-method="onListenerRegistered" unbind-method="onListenerUnregistered" />
     </onmsgi:list>
+
```
Comment (j-white) [inline opennms-dao/src/main/resources/META-INF/opennms/applicationContext-shared.xml:292]: Unnecessary change.

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc [hunk 12]
```diff
@@ -10,7 +10,7 @@ With _telemetryd_, operators can define listeners supporting different protocols
 image::telemetryd/telemetryd-overview.png[]
 
 The configuration is split in two parts.
-_Listeners_ and attached _Parsers_ on one side, receive telemetry data transported over a specific protocol and parse the protocol according to its specification.
+_Listeners_ and attached _Parsers_ or _Connectors_ on one side, receive telemetry data transported over a specific protocol and parse the protocol according to its specification.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc:13]:
```suggestion
_Listeners_ and attached _Parsers_ or _Connectors_ on one side receive telemetry data transported over a specific protocol and parse the data according to the protocol's specification.
```

I believe it is the data that is parsed, not the protocol, hence my suggested edit.

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc [hunk 13]
```diff
@@ -28,6 +28,15 @@ _Parsers_ and _Listeners_ must work together closely.
 Therefore, they cannot be combined in any random way.
 Which _Parser_ is supported by which _Listener_ is described in the documentation of each respective _Parser_.
 
+=== Connectors
+
+_Connectors_ initiate a connection and maintain a session with agent. A new connector is created for every agent it is responsible for connecting to.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc:33]:
```suggestion
_Connectors_ initiate a connection and maintain a session with an agent. A new connector is created for every agent it is responsible for connecting to.
```

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc [hunk 14]
```diff
@@ -28,6 +28,15 @@ _Parsers_ and _Listeners_ must work together closely.
...
+Agent here refers to an IP-Service which is a named service attached to an IP interface on a specific node.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc:34]:
```suggestion
Agent here refers to an IP service that is a named service attached to an IP interface on a specific node.
```

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc [hunk 15]
```diff
@@ -28,6 +28,15 @@ _Parsers_ and _Listeners_ must work together closely.
...
+Connectors don't use parsers and dispatch messages directly to queue.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/introduction.adoc:35]:
```suggestion
Connectors don't use parsers, and dispatch messages directly to a queue.
```

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc [hunk 16]
```diff
@@ -51,7 +51,6 @@ ${OPENNMS_HOME}bin/send-event.pl -p 'daemonName Telemetryd' uei.opennms.org/inte
 
 OpenConfig Connector initiates a connection via gRPC and subscribe to one or more OpenConfig data paths (strings) and then forwards the data to queue.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc:52]:
```suggestion
OpenConfig Connector initiates a connection via gRPC and subscribes to one or more OpenConfig data paths (strings) and then forwards the data to a queue.
```

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc [hunk 17]
```diff
@@ -75,9 +74,13 @@ OpenConfig Connector initiates a connection via gRPC and subscribe to one or mor
 
 ----
 
-
 ====== Packages
-Connector can have multiple packages. A package consists of filter and a list of parameters specific to interfaces that match the filter.
+At least one package must be present for the connector to function.
+Multiple packages can be used to customize the parameters subsets of nodes.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc:79]:
```suggestion
Use multiple packages to customize the parameter's subsets of nodes.
```

Do the subsets of nodes belong to the parameter?

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc [hunk 18]
```diff
@@ -75,9 +74,13 @@ OpenConfig Connector initiates a connection via gRPC and subscribe to one or mor
...
+
+The `filter` element is optional, if missing, all services with the given name will be considered.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc:81]:
```suggestion
The `filter` element is optional; if missing, all services with the given name will be considered.
```

File: opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc [hunk 19]
```diff
@@ -75,9 +74,13 @@ OpenConfig Connector initiates a connection via gRPC and subscribe to one or mor
...
+
+Parameters are passed to the connector and are interpolated for node/interface & service level meta-data.
```
Comment (Bonrob2) [inline opennms-doc/guide-admin/src/asciidoc/text/telemetryd/protocols/openconfig.adoc:83]:
```suggestion
Parameters are passed to the connector and are interpolated for node/interface and service-level metadata.
```

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement

**Comment by j-white** [comment]
Hey Ulf, approve jesse@opennms.com

**Comment by j-white** [comment]
Good to merge from my perspective.

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](https://wiki.opennms.org/wiki/images/6/6b/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit https://wiki.opennms.org/wiki/Contributor_Agreement
